### DISCUSSION TITLE
Audio interface channel names in setup dialog

### DISCUSSION THREAD

**Discussion by AndersGoran** [jamulussoftware/jamulus#3400] [General]
I was hesitant if I should file this under bugs or whatever, but I see an interesting thing when I compare pre-built vs self-compiled Jamulus clients.

I use macOS (currently 12.7.6, M1 laptop). The audio interface I most often use is an old MOTU 828mk3, the firewire-only version. Not officially supported by MOTU but it actually works very well under Monterey once properly setup.

Whenever I run an official release of Jamulus, or for that matter, the build artifact from the recent --midictrlch PR, the audio setup dialog looks like this:
<img width="571" alt="Official client" src="https://github.com/user-attachments/assets/86c5b21a-d757-4beb-b8e6-bb5d26d3fde6">

I.e., only generic names for all input and output channels are shown. It shows the correct number of channels, depending on whether I enable or disable ADAT and so on. 

If I run a client that I compiled myself, I get this:
<img width="555" alt="Locally compiled client" src="https://github.com/user-attachments/assets/59a194e0-fee0-4642-a15b-96795e5b7402">

It is of course much nicer to see the channel names that I've configured in MOTU's setup app for the interface. I don't expect anyone to have an explanation or fix, but I find it interesting and I'll do some digging to see if I can understand it.

**Comment by ann0see** [comment]
Might be a different Qt version.

**Reply by AndersGoran** [reply]
I have been using Qt from Homebrew to compile Jamulus, since that is what `COMPILING.md` recommends. Is that still the recommended Qt distribution to use? 

Both 5.15 and 6.7.0 can be used to build a working client, and they both behave the same with regard to channel names. The names are certainly not hardcoded into the client; if I adjust a channel name and restart Jamulus, the change is reflected.

**Reply by ann0see** [reply]
That's strange. I don't own a real multi channel device so I cannot say much about it. 
Which Xcode version are you using?

**Reply by AndersGoran** [reply]
Xcode is 13.1, compiler is
```
$ clang -v
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I'm not that savvy with Xcode for debugging, but currently trying to understand where these names are coming from. From Qt, I suppose? 

It's not a huge problem, just an inconvenience that I need to count channels to pick the right ones when their names aren't shown. E.g. I usually want to send Jamulus output to Phones (11/12, I have somewhat memorised) and often but not always, I want to capture my DAW's output so I can use plugins + stuff on my instrument, so I want inputs 13, 14 ("Return", a kind of loopback input that mirrors a selectable output pair). But I tend to forget, was it 11/12 for input and 13/14 for output, or vice versa? 

One thing I'll do is bring my build to another machine, to see if it makes any difference that it wasn't built locally.

If anyone else is reading this, and you are also using multi-channel audio hardware, do you see meaningful channel names or just generic names with official builds?

**Reply by ann0see** [reply]
We currently use Xcode 15.4 so that's a newer SDK for the official builds. 
The names should not come from Qt, but sometimes different Qt versions lead to odd behavior.

**Reply by ann0see** [reply]
I'd start looking at https://github.com/jamulussoftware/jamulus/blob/main/src/sound/coreaudio-mac/sound.cpp

**Reply by AndersGoran** [reply]
Tonight's experiment was to build the client in a quite clean Sonoma system.

Xcode 15.4 with `MacOSX14.5.sdk` and `Homebrew Qt 6.7.2`, everything freshly installed, no old cruft laying around. 

It is a lot like what the Github autobuild process seems to use. I manually changed `-mmacosx-version-min=14.0` found in `Makefile.{Debug,Release}` into `-mmacosx-version-min=11.0` to match autobuild even better. 

Outcome: my builds (both debug and release) still show proper channel names. Official 3.11 client shows generic names "Channel XX", i.e. just as before. Puzzling!

...

BUUUT: finally something!! So far, I've only done `make release`or `make debug` and the resulting binaries from those show useful channel names. The Jamulus.app I get from `mac/deploy_mac.sh` **_does not_** show proper channel names! So that one has a lot of extra stuff bundled into it... why on earth can that one not get the same information when it inquires about the audio interface???

**Reply by ann0see** [reply]
That's maybe due to different build time flags? 
I think we use make instead of xcodebuild for the release.

**Reply by AndersGoran** [reply]
Yes, you do:
<img width="921" alt="build command" src="https://github.com/user-attachments/assets/ff34f0e5-d2bd-490c-b915-7f5a2c05298c">

And so do I. 

I have to compare very closely what happens in the auto build, and also after the `make` command has finished, when it bundles Qt dependencies into the app that goes onto the disk image.

**Reply by ann0see** [reply]
The macdeployqt script does some strange stuff to the app. 
Also https://github.com/jamulussoftware/jamulus/issues/3078 seems like we do something wrong conceptually.

**Reply by AndersGoran** [reply]
Takes me back to where I got into this whole thing, wanting to learn to compile the client myself... 

I think `macdeployqt` is meant to bundle the Qt libs into the app so end users don't have to have Homebrew + Qt installed. My _guess_ is that a locally compiled app (in my case built with `make` rather than `xcodebuild`) runs fine without Qt libs bundled because they're already there on the machine and somehow found automatically.

**Reply by ann0see** [reply]
The question is, could we build without macdeployqt but with signing etc.

**Reply by ann0see** [reply]
Potentially we want to check out the differences between Info-make.plist and Info-Xcode.plist also.

**Reply by ann0see** [reply]
Does this issue also occur on beta1 of 3.11.0, unsigned version?

**Reply by AndersGoran** [reply]
Will check the beta tomorrow. Time for bed now.

**Reply by AndersGoran** [reply]
I suppose you mean [jamulus_3.11.0beta_macUnsigned.dmg](https://github.com/jamulussoftware/jamulus/releases/download/r3_11_0beta1/jamulus_3.11.0beta_macUnsigned.dmg)? Yes, generic names with that one.

**Reply by ann0see** [reply]
Good to know. Otherwise it could have been related to signing

**Reply by AndersGoran** [reply]
Yep, I don't think the signing has to do with it. The file you indicated, `coreaudio-mac/sound.cpp`, it has this interesting section (or actually two of them, one for inputs, one for outputs):

```
AudioObjectGetPropertyData ( audioInputDevice[iDriverIdx], &stPropertyAddress, 0, NULL, &iPropertySize, &sPropertyStringValue );

// convert string
const bool bConvOK = ConvertCFStringToQString ( sPropertyStringValue, sChannelNamesInput[iCurInCH] );

// add the "[n]:" at the beginning as is in the Audio-Midi-Setup
if ( !bConvOK || ( iPropertySize == 0 ) )
{
    // use a default name in case there was an error or the name is empty
    sChannelNamesInput[iCurInCH] = QString ( "%1: Channel %1" ).arg ( iCurInCH + 1 );
}
else
{
    sChannelNamesInput[iCurInCH].prepend ( QString ( "%1: " ).arg ( iCurInCH + 1 ) );
}
```

This is clearly where the channel names are constructed, so I need to figure out which one of `!bConvOK || ( iPropertySize == 0 )` is happening, leading to "default" names.

**Reply by AndersGoran** [reply]
```
// get the channel names of the output device
  for ( int iCurOutCH = 0; iCurOutCH < iNumOutChan; iCurOutCH++ )
  {
      CFStringRef sPropertyStringValue = NULL;
      stPropertyAddress.mSelector = kAudioObjectPropertyElementName;
      stPropertyAddress.mElement  = iCurOutCH + 1;
      stPropertyAddress.mScope    = kAudioObjectPropertyScopeOutput;
      iPropertySize               = sizeof ( CFStringRef );

      AudioObjectGetPropertyData ( audioOutputDevice[iDriverIdx], &stPropertyAddress, 0, NULL, &iPropertySize, &sPropertyStringValue );
```

A client that has been run through `macdeployqt` gets zero-length strings for the audio channel names in `sPropertyStringValue`. The string conversion in `ConvertCFStringToQString` fails because it has no characters to work on, hence the default names.

**Reply by ann0see** [reply]
This sounds like a sandbox issue still.

**Reply by AndersGoran** [reply]
OK... this is so strange, I've no idea what to think. Maybe the code signature has something to do with this after all. 

I have figured out that if I take the official macOS client, install it from the official DMG and _re-sign it_, like this:
```
$ codesign --force --sign - /Applications/Jamulus.app
/Applications/Jamulus.app: replacing existing signature
```
so that it gets some kind of ad-hoc signature instead... and now I have to right-click and choose Open to launch it, as this supposedly breaks the notarization... 

**lo and behold, I now see the sound interface's real channel names in Settings!**  

At first I had the `--deep` option too, so I was re-signing everything in the app bundle, but that's not necessary, it turns out.
